packages/sdk/src/transactions/PSBTBuilder.ts

```diff
@@ -0,0 +1,468 @@
+      this.psbt.addOutput({
+        address: this.changeAddress || this.address,
+        value: this.changeAmount,
+      });
+    }
+  }
+
+  private calculateOutputAmount() {
+    this.outputAmount = Math.floor(
+      this.outputs.reduce((acc, curr) => acc + curr.value, 0) +
+        this.injectableOutputs.reduce((acc, curr) => acc + curr.sats, 0),
+    );
+
+    this.validateOutputAmount();
+  }
+
+  private async calculateChangeAmount() {
+    if (!this.autoAdjustment) {
+      return;
+    }
+
+    this.changeAmount = Math.floor(
+      this.inputAmount - this.outputAmount - this.fee,
+    );
+
+    await this.isNegativeChange();
+  }
+
+  private async isNegativeChange() {
+    if (this.changeAmount >= 0) {
+      return;
+    }
+
+    await this.prepare();
+    if (this.noMoreUTXOS) {
+      throw new Error(
+        `Insufficient balance. Decrease the output amount by ${
+          this.changeAmount * -1
+        } sats`,
+      );
+    }
+  }
+
+  private getRetrievedUTXOsValue() {
+    return this.utxos.reduce((acc, utxo) => acc + utxo.sats, 0);
+  }
+
+  private getReservedUTXOs() {
+    return this.utxos.map((utxo) =>
+      generateTxUniqueIdentifier(utxo.txid, utxo.n),
+    );
+  }
+
+  private getUTXOAmountToRequestFromChangeAmount() {
+    if (this.changeAmount < 0) {
+      return Math.abs(this.changeAmount);
+    }
+
+    return this.outputAmount - this.getRetrievedUTXOsValue();
+  }
+
+  private async retrieveUTXOs(address?: string, amount?: number) {
+    if (!this.autoAdjustment && !address) {
+      return;
+    }
+
+    const amountToRequest =
+      amount && amount > 0
+        ? amount
+        : this.getUTXOAmountToRequestFromChangeAmount();
+    if (
+      (amount && this.getRetrievedUTXOsValue() >= amount) ||
+      amountToRequest <= 0
+    ) {
+      return;
+    }
+
+    const utxos = await this.datasource.getSpendables({
+      address: address || this.address,
+      value: convertSatoshisToBTC(amountToRequest),
+      filter: this.getReservedUTXOs(),
+    });
+
+    this.noMoreUTXOS = utxos.length === 0;
+
+    this.utxos.push(...utxos);
+  }
+
+  protected async retrieveSelectedUTXOs(address: string, amount: number) {
+    await this.retrieveUTXOs(address, amount);
+    const selectedUTXOs = this.utxos.find((utxo) => utxo.sats >= amount);
+
+    this.utxos = selectedUTXOs ? [selectedUTXOs] : [];
+
+    return this.utxos;
+  }
+
+  private async prepareInputs() {
+    if (!this.autoAdjustment) {
+      return undefined;
+    }
+
+    const promises: Promise<InputType>[] = [];
+
+    this.utxos.forEach((utxo) => {
+      if (
+        this.usedUTXOs.includes(generateTxUniqueIdentifier(utxo.txid, utxo.n))
+      ) {
+        return;
+      }
+
+      this.inputAmount += utxo.sats;
+      const promise = processInput({
+        utxo,
+        pubKey: this.publicKey,
+        network: this.network,
+        datasource: this.datasource,
+      }); // TODO: add sigHashType
```

Comment:
@iamcrazycoder is `sigHashType` being used only in `InstantTradeSellerTxBuilder`? what will happen after this is added here?

Reply:
`FeeEstimator` doesn't consider `sigHashType` atm, and this `// TODO` has been added to cover that edge case.